libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationV2Router.kt

```diff
@@ -0,0 +1,39 @@
+package com.mapbox.navigation.base.route
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.navigation.base.CurrentIndicesSnapshot
+
+/**
+ * Extends [NavigationRouter] to also provide ability for refreshing routes partially
+ * using indices stored in [CurrentIndicesSnapshot].
+ */
+interface NavigationV2Router : NavigationRouter {
+
+    @Deprecated(
+        "Use getRouteRefresh(" +
+            "NavigationRoute, " +
+            "CurrentIndicesSnapshot, " +
+            "NavigationRouterRefreshCallback" +
+            ")"
+    )
+    override fun getRouteRefresh(
+        route: NavigationRoute,
+        legIndex: Int,
+        callback: NavigationRouterRefreshCallback
+    ): Long
```

Comment:
Users will have to implement this method anyway if they provide a custom router, won't they? Maybe the SDK should provide a default implementation which trows an exception?

Reply:
Oh, I've just realised that we have the same problem with `NavigationRouter` 😱 
Customers have to override methods from `Router` despite the fact that the SDK doesn't use them. @LukasPaczos , was it done on purpose or we just didn't notice it?

Reply:
Yes. And, as I mentioned [here](https://github.com/mapbox/mapbox-navigation-android/pull/6115#issuecomment-1218017285), the module where the implementation is has to modify its `build.gradle` so we can't actually provide the default implementation all by ourselves. I also find it weird but I double-checked. 
Other not-so-pretty options are:
1) abstract class
2) interface in java

Reply:
> I mentioned https://github.com/mapbox/mapbox-navigation-android/pull/6115#issuecomment-1218017285, the module where the implementation is has to modify its build.gradle so we can't actually provide the default implementation all by ourselves

But it woks like this only for java? Can we make at least kotlin's users life a bit easier?

Reply:
> interface in java

I think it's a nice option if we decide to provide default implementation. But let's wait for @LukasPaczos to decide if we need default implementation?

Reply:
>  default implementation which trows an exception

Looking at it from the perspective of the consumer of the interface, a piece of code that implements it, it has to remain compatible with `Router` and `NavigationRouter` interfaces. It doesn't "know" that one of the methods will not be used anymore, it just provides the implementation and has no idea who will call it. The contract that the method will not be called (and could have an exception-based default impl) would only rely on documentation, and I don't see this being a good practice.

If we want to provide a default impl for Kotlin users, I think it might be best to re-wire the old function into the new one, and that's all.

Reply:
> If we want to provide a default impl for Kotlin users, I think it might be best to re-wire the old function into the new one, and that's all.

What do you mean? It would break API.

Reply:
`NavigationRouterV2` is not released yet, so what would we break?

I'm thinking about:
```
interface NavigationRouterV2 : NavigationRouter {
    override fun getRouteRefresh(
        route: NavigationRoute,
        legIndex: Int,
        callback: NavigationRouterRefreshCallback
    ): Long {
        return getRouteRefresh(
            route, CurrentIndices(legIndex, {find route index of a first point in the leg}, 0), callback
        )
    }
    ...
}
```

That said, overwriting a router is a very minor use-case (if anything), the investment might not make a ton of sense and we probably should just leave the methods as they are.

Reply:
I didn't notice us moving the interface to `internal`, then this conversation is all invalid, we can do whatever we want and need 🙃

Reply:
> NavigationRouterV2 is not released yet, so what would we break?

We would still require java users to implement the method unless **they** specify in **their** `build.gradle` specific compiler arguments. More on that: https://github.com/mapbox/mapbox-navigation-android/pull/6190#discussion_r949981502

> I didn't notice us moving the interface to internal, then this conversation is all invalid, we can do whatever we want and need 🙃

Yes, for now there's no problem, just want to make sure we're on the same page. Maybe I'm missing something obvious.

Reply:
> We would still require java users to implement the method unless they specify in their build.gradle specific compiler arguments.

We're not requiring anyone to implement and use `NavigationRouterV2`, they can remain on `NavigationRouter` or just `Router`. That's why I don't see any breaking changes being introduced here.